### ISSUE TITLE
Migrate Consultation Response Forms to Asset Manager

### ISSUE THREAD

**Issue by chrislo** [alphagov/asset-manager#297]
These are the forms added to consultations, for example this [response form](https://www.gov.uk/government/uploads/system/uploads/consultation_response_form_data/file/622/revised-draft-airports-nps-response-form.docx) linked to under "Complete a response form" on [this Heathrow expansion consultation](https://www.gov.uk/government/consultations/heathrow-expansion-revised-draft-airports-national-policy-statement). 

- [x] Change `ConsultationResponseFormUploader` to use `:asset_manager_and_quarantined_file_storage` storage engine (send uploads to Asset Manager as well as save on NFS)
- [x] Migrate existing consultation response forms under `system/uploads/consultation_response_form_data/file` to Asset Manager
- [x] Ensure consultation response forms are served on assets.publishing instead of gov.uk
  * https://github.com/alphagov/whitehall/pull/3634
- [x] Redirect consultation response forms from gov.uk to assets.publishing
  * https://github.com/alphagov/govuk-puppet/pull/6890
  * https://github.com/alphagov/whitehall/pull/3627
- [x] Change `nginx` to serve requests to these paths from Asset Manager rather than Whitehall
  * https://github.com/alphagov/govuk-puppet/pull/6776
- [x] Deploy changes in https://github.com/alphagov/govuk-puppet/pull/6776 to production
- [x] Change `ConsultationResponseFormUploader` to use `:asset_manager` storage engine (stop saving new uploads to NFS)
- [x] Merge https://github.com/alphagov/whitehall/pull/3682
- [x] Test the effect of https://github.com/alphagov/whitehall/pull/3682 in integration
- [x] Delete existing data from NFS (run `asset_manager:remove_consultation_response_form_data_file`)

**Comment by chrislo** [comment]
@emmabeynon ran the rake task `asset_manager:migrate_assets[system/uploads/consultation_response_form_data/file/]` against production today (`2017-11-13T15:38:47.537272`). We observed that 732 assets were enqueued in Whitehall on the `asset-migration` queue. The queue was empty by 15:40.

**Comment by chrislo** [comment]
I've opened https://github.com/alphagov/govuk-puppet/pull/6776 to make the change to `nginx` to serve these assets from Asset Manager.

**Comment by chrisroos** [comment]
I've just been looking at merging but, while doing so, have noticed that consultation response forms appear to be served from gov.uk instead of assets.publishing. I'm going to investigate that further as it might mean we have some other changes to make before we can merge the nginx changes.

**Comment by chrisroos** [comment]
I've opened https://github.com/alphagov/govuk-puppet/pull/6890 so that we pass the host (gov.uk or assets.publishing.service.gov.uk) to Whitehall. This will allow us to redirect requests for assets that are made to gov.uk.

**Comment by chrisroos** [comment]
@danielroseman has suggested that we should be able to add this redirect (from www.gov.uk to assets.publishing.service.gov.uk) in the router so I'm going to investigate that instead of doing it in Whitehall.

For reference, my unfinished implementation of this in Whitehall is in https://github.com/alphagov/whitehall/tree/redirect-asset-requests-to-asset-host.

**Comment by chrisroos** [comment]
I've created issue #348 to capture the work required to redirect assets to the asset host.

**Comment by chrisroos** [comment]
I've opened https://github.com/alphagov/whitehall/pull/3634 to use the asset host in the links to consultation response forms that we save to the Content Store.

**Comment by chrisroos** [comment]
https://github.com/alphagov/whitehall/pull/3634 has been deployed to production which means that we're now using the asset host when storing links to the consultation response forms in the Content Store.

**Comment by chrisroos** [comment]
The redirection of assets to the asset host has now been merged to production (see issue #348 for more info).

**Comment by chrisroos** [comment]
I tried to merge https://github.com/alphagov/govuk-puppet/pull/6776 to serve consultation response forms from Asset Manager but ran into a discrepancy in the number of assets on the Whitehall filesystem versus Asset Manager database. I've added a comment to that PR and have created #368 to prompt us to address the discrepancy.

**Comment by chrislo** [comment]
We've migrated the two missing files in https://github.com/alphagov/asset-manager/issues/382 so we can carry on with this issue now.

**Comment by chrisroos** [comment]
I've just merged https://github.com/alphagov/govuk-puppet/pull/6776 and will check the effect in integration as soon as it's been automatically deployed/run.

**Comment by chrisroos** [comment]
I tried to compare the effect of https://github.com/alphagov/govuk-puppet/pull/6776 in integration but it hasn't actually had any effect because we're missing some asset_routes from hieradata_aws. I've opened https://github.com/alphagov/govuk-puppet/pull/7069 to address this.

**Comment by chrisroos** [comment]
https://github.com/alphagov/govuk-puppet/pull/7069 has been merged to integration and I can see that the nginx config should now be routing requests for consultation response form assets to Asset Manager instead of Whitehall.

I requested an asset before and after switching the routing to Asset Manager:

## When routing consultation response form requests to Whitehall

```
$ curl -v "https://assets-origin.integration.publishing.service.gov.uk/government/uploads/system/uploads/consultation_response_form_data/file/622/revised-draft-airports-nps-response-form.docx" > /dev/null

* Connection state changed (MAX_CONCURRENT_STREAMS updated)!
< HTTP/2 200
< date: Thu, 11 Jan 2018 12:52:31 GMT
< content-type: application/vnd.openxmlformats-officedocument.wordprocessingml.document
< content-length: 77666
< server: nginx
< accept-ranges: bytes
< cache-control: max-age=14400, public
< content-disposition: inline; filename="revised-draft-airports-nps-response-form.docx"
< etag: "59edd58b-12f62"
< last-modified: Mon, 23 Oct 2017 11:42:03 GMT
< x-frame-options: SAMEORIGIN
< access-control-allow-origin: *
< access-control-allow-methods: GET, OPTIONS
< access-control-allow-headers: origin, authorization
```

## When routing consultation response form requests to Asset Manager
```
$ curl -v "https://assets-origin.integration.publishing.service.gov.uk/government/uploads/system/uploads/consultation_response_form_data/file/622/revised-draft-airports-nps-response-form.docx" > /dev/null

* Connection state changed (MAX_CONCURRENT_STREAMS updated)!
< HTTP/2 200
< date: Fri, 12 Jan 2018 12:50:41 GMT
< content-type: application/octet-stream
< content-length: 77666
< server: nginx
< accept-ranges: bytes
< cache-control: max-age=14400, public
< content-disposition: inline; filename="revised-draft-airports-nps-response-form.docx"
< etag: "59edd54e-12f62"
< last-modified: Mon, 23 Oct 2017 11:41:02 GMT
< strict-transport-security: max-age=31536000
< x-frame-options: SAMEORIGIN
< access-control-allow-origin: *
< access-control-allow-methods: GET, OPTIONS
< access-control-allow-headers: origin, authorization
```

Note the differences in the `content-type`, `last-modified` and `strict-transport-security` headers.

I've created https://github.com/alphagov/asset-manager/issues/408 to investigate the discrepancy in `content-type`.

The small difference in `last-modified` will be because of the difference in time between the file being saved to Whitehall and it being saved to Asset Manager. I'm not worried about this.

I'm not worried about the difference in the `strict-transport-security` header.

**Comment by chrisroos** [comment]
I need to wait until these Puppet changes have been deployed to production before I can make the other changes in the todo list in the description of the issue.

**Comment by chrisroos** [comment]
The changes in https://github.com/alphagov/govuk-puppet/pull/6776 were deployed to production when release_19877 was deployed on Fri 12 Jan at 11:11.

**Comment by chrisroos** [comment]
Unfortunately, a separate issue is currently preventing consultation response forms from being accessed. See issue #412.

**Comment by chrisroos** [comment]
Issue #412 has now been resolved and I've confirmed that consultation response forms are being served by Asset Manager. I used curl to make the following request and used Kibana to confirm that the request was routed to Asset Manager:

```
$ curl -v "https://assets-origin.publishing.service.gov.uk/government/uploads/system/uploads/consultation_response_form_data/file/622/revised-draft-airports-nps-response-form.docx?CJR$RANDOM" > /dev/null

> GET /government/uploads/system/uploads/consultation_response_form_data/file/622/revised-draft-airports-nps-response-form.docx?CJR1818 HTTP/1.1
> Host: assets-origin.publishing.service.gov.uk
> User-Agent: curl/7.54.0
> Accept: */*

< HTTP/1.1 200 OK
<snipped>
```

I searched Kibana for the `CJR1818` querystring and saw the following entries:

```
January 15th 2018, 15:01:26.716	 - 	 - 	asset-manager
January 15th 2018, 15:01:26.000	 - 	 - 	static.publishing.service.gov.uk-json.event.access
January 15th 2018, 15:01:26.000	 - 	 - 	asset-manager.publishing.service.gov.uk-json.event.access
January 15th 2018, 15:01:26.000	 - 	 - 	asset-manager.publishing.service.gov.uk-json.event.access
January 15th 2018, 15:01:26.000	 - 	 - 	static.publishing.service.gov.uk-json.event.access
January 15th 2018, 15:01:26.000	 - 	 - 	assets-origin.publishing.service.gov.uk-json.event.access
```

**Comment by chrisroos** [comment]
I've opened https://github.com/alphagov/whitehall/pull/3682 to stop saving assets to NFS.

**Comment by chrisroos** [comment]
The changes in https://github.com/alphagov/whitehall/pull/3682 were deployed to integration about 4 days ago.

I've just tested the effect of this change in integration. I created a new consultation and attached a consultation response form. The form was available at https://assets-origin.integration.publishing.service.gov.uk/government/uploads/system/uploads/consultation_response_form_data/file/637/headshot.jpg.pdf. I can see that no consultation_response_form_data directory has been created in /data/uploads/whitehall/clean/system/uploads so I'm happy that we're no longer saving consultation response form files to the Whitehall directory on NFS.

**Comment by chrislo** [comment]
I ran `asset_manager:remove_consultation_response_form_data_file` [in integration](https://deploy.integration.publishing.service.gov.uk/job/run-rake-task/1/console) and verified that it removed all the files and directories under `/data/uploads/whitehall/clean/system/uploads/consultation_response_form_data`.

**Comment by chrislo** [comment]
We ran this in production
